From 98a4b0fc0ee290c30bdd57434c5572f1bd54b11d Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Mon, 02 Nov 2020 19:21:39 +0200
Subject: [PATCH] [SW-17915] ppboot: module_pci: remove pcie_en reg

Remove usage of mmPREBOOT_PCIE_EN register.
Instead, before PCIe enable, we first check the
reset cause (SW_ALL_RESET, watchdog, cold reset, etc.),
and decide accordingly if PCIe should be initialized or not.
Note that in this patch we still only imitate the previous behaviour
(with mmPREBOOT_PCIE_EN and the MAGIC words), but instead of using
PCIE_EN register, we decide if PCIe should be initialized or not
according to reset cause.

Change-Id: I8d5dfb294261cfbf670f2b1d84b4900acf5d0c33
Signed-off-by: Ran Shalit <rshalit@habana.ai>
---

diff --git a/gaudi/common.c b/gaudi/common.c
index c1a9a63..7cfad43 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -116,7 +116,7 @@
 	return (pci_dev_id << 16) | PCI_VENDOR_ID_HABANALABS;
 }
 
-static void preboot_init_pcie(void)
+void preboot_init_pcie(void)
 {
 	u32 ven_id_dev_id = preboot_get_vendor_id_device_id();
 
@@ -153,15 +153,6 @@
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
 }
 
-u32 preboot_clr_flops_reset_cause(void)
-{
-	u32 reset_cause;
-
-	reset_cause = regmap_readl(mmPREBOOT_PCIE_EN);
-	regmap_writel(0, mmPREBOOT_PCIE_EN);
-	return reset_cause;
-}
-
 void ppboot_pcie_enable_access(void)
 {
 	u32 reg;
@@ -173,26 +164,6 @@
 	reg = regmap_readl(mmPCIE_DBI_STATUS_COMMAND_REG);
 	reg |= BIT(PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT);
 	regmap_writel(reg, mmPCIE_DBI_STATUS_COMMAND_REG);
-}
-
-/*
- * This logic copied from BTL PCIe section (#4)
- * In BTL it is called when DIP switch for PCIe enable is set
- * Now we wish to do this logic also here as a measure of security
- * It means that in new PCB's version by default PCIe won't be enabled
- * So without proper FW device won't be accessible through PCIe
- * Hence device is more secured
- */
-static void preboot_pcie_enable(void)
-{
-	u32 reset_cause;
-
-	/* During hard reset triggered by LKD keep PCIe intact */
-	reset_cause = preboot_clr_flops_reset_cause();
-	if (reset_cause == LKD_HARD_RESET_MAGIC)
-		return;
-
-	preboot_init_pcie();
 }
 
 void preboot_pcie_wait_for_link(void)
diff --git a/goya2/common.c b/goya2/common.c
index 6c1222b..8691254 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -217,7 +217,7 @@
 	preboot_init_pcie_phy_fw_update();
 }
 
-static void preboot_init_pcie(void)
+void preboot_init_pcie(void)
 {
 	preboot_init_pcie_phy();
 
@@ -243,27 +243,6 @@
 
 	/* Enable and Mask LTSSM (erratum 102) */
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-}
-
-u32 preboot_clr_flops_reset_cause(void)
-{
-	u32 reset_cause;
-
-	reset_cause = regmap_readl(mmPREBOOT_PCIE_EN);
-	regmap_writel(0, mmPREBOOT_PCIE_EN);
-	return reset_cause;
-}
-
-static void preboot_pcie_enable(void)
-{
-	u32 reset_cause;
-
-	/* During hard reset triggered by LKD keep PCIe intact */
-	reset_cause = preboot_clr_flops_reset_cause();
-	if (reset_cause == LKD_HARD_RESET_MAGIC)
-		return;
-
-	preboot_init_pcie();
 }
 
 void preboot_pcie_wait_for_link(void)
diff --git a/goya2_arc/common.c b/goya2_arc/common.c
index 6ba72f5..dcdb91c 100644
--- a/goya2_arc/common.c
+++ b/goya2_arc/common.c
@@ -124,7 +124,7 @@
 
 }
 
-static void preboot_init_pcie(void)
+void preboot_init_pcie(void)
 {
 
 	preboot_init_pcie_phy();
@@ -145,19 +145,6 @@
 
 	/* Enable and Mask LTSSM (erratum 102) */
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
-}
-
-static void preboot_pcie_enable(void)
-{
-	u32 reset_cause;
-
-	/* During hard reset triggered by LKD keep PCIe intact */
-	reset_cause = regmap_readl(mmPREBOOT_PCIE_EN);
-	regmap_writel(0, mmPREBOOT_PCIE_EN);
-	if (reset_cause == LKD_HARD_RESET_MAGIC)
-		return;
-
-	preboot_init_pcie();
 }
 
 void preboot_pcie_wait_for_link(void)
diff --git a/include/preboot.h b/include/preboot.h
index bac3ff0..8ded23e 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -115,6 +115,7 @@
 void preboot_putc_virt_uart(const char c);
 void preboot_sync_btl(void);
 void preboot_led_on(void);
+void preboot_pcie_enable(void);
 
 void ppboot_cp_preboot(u64 loadaddr, u32 offset, u32 size, char *name);
 void image_start(u64 loadaddr);
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index ac70971..4c111b6 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -104,6 +104,34 @@
 	return false;
 }
 
+__weak void preboot_init_pcie(void) {}
+
+void preboot_pcie_enable(void)
+{
+	u32 rst_src;
+
+	rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+	switch (rst_src) {
+	case HL_COLD_RST:
+	case HL_PRSTN_RST:
+		/* PCIe is reset, so must initialize PCIe */
+		break;
+	case HL_MANUAL_RST:
+	case HL_WD_RST:
+	case HL_SW_ALL_RST:
+		/*
+		 * Currently, keep using the same flow as done before,
+		 * i.e. PCIe should already be initialized by default.
+		 * During hard reset triggered by LKD keep PCIe intact.
+		 */
+		return;
+	default:
+		break;
+	}
+
+	preboot_init_pcie();
+}
+
 static void rst_src_dump(void)
 {
 	unsigned int rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
@@ -153,14 +181,12 @@
 	return false;
 }
 
-__weak u32 preboot_clr_flops_reset_cause(void) { return 0; }
 __weak void ppboot_log_waiting_for_pcie(void) {}
 __weak void ppboot_pcie_enable_access(void) {}
 
 static void ppboot_pcie_init(void)
 {
 	if (hl_pcie_link_up()) {
-		preboot_clr_flops_reset_cause();
 		ppboot_pcie_enable_access();
 		preboot_puts("PCIe link is already UP!\n");
 		return;
